Silent Payments: send and receive

josibake opened the pull request:
## For reviewers

In an attempt to make reviewing a bit more sane, I'm breaking this up into a few smaller PRs, but will keep this one open as the parent PR and keep it rebased on the child PRs. The main purpose of having this PR is to track progress on child PRs and also have an easy way to compile `bitcoind` with both send and receive support for testing. Additionally, I'll be adding more functional tests to this PR since it's much easier to test when `bitcoind` can both send and receive.

## PRs

- [ ] https://github.com/bitcoin/bitcoin/pull/28122
  - Implements the logic from BIP352 without any wallet code. This PR adds the necessary cryptographic functions and implements the logic needed for sending and scanning. This PR also includes the test vectors from the BIP as unit tests. Both the send and receive PRs have this as a dependency. In terms of priority, this PR should be reviewed first
- [ ] https://github.com/bitcoin/bitcoin/pull/28201
  - Implements sending in the Bitcoin Core wallet. This PR allows a wallet to send to a silent payment address, regardless of whether or not the wallet can receive silent payments
  - Ready for review, but marked as a draft until dependencies are merged
- [ ] https://github.com/bitcoin/bitcoin/pull/28202
  - Implements receiving in the Bitcoin Core wallet. This PR allows a wallet to generate silent payment addresses and scan for silent payments, regardless of whether or not the wallet can send to a silent payment address
  - Ready for review but marked as a draft until dependencies are merged

For the silent payments specification, see https://github.com/bitcoin/bips/pull/1458

## Overall

This PR implements the full silent payments scheme: sending and receiving. The following items are not covered in this PR and are intended for follow-up PRs:

* Adding labels for the receiver wallet
* Full RPC coverage (only `sendtoaddress` and `sendmany` are covered in this PR)
* Light client support (vending the tweak data per block, either in an index or to serve to indexer, such as electrum server)
* Add benchmarks to validate that there are no DoS concerns for doing silent payment verification for transactions in the mempool
* External signer support (dependent on hardware wallets supporting silent payments)
* More unit / functional test coverage

## Major changes

This PR is a continuation of the work done in https://github.com/bitcoin/bitcoin/pull/24897. Below is a summary of the major changes:

* Remove labels
  * The original draft included labels, but this has been deferred for a later PR. Labels are not necessary for sending and receiving and there are still some open questions on how best to implement them in Bitcoin Core. Labels can also be added at any point by the receiver without requiring any changes from the sender
* Remove indexes
  * In the original draft, indexes were used when scanning for silent payments and when doing wallet rescans. This has been removed in favor of using `rev*.dat` files for rescanning. It may make sense to add an index in the future, but for the purpose of vending tweak data to light clients, which is still an open question
* Update to implement the most current version of BIP352
  * Since the original draft, there have been a few changes in the BIP which are reflected in the current PR. Most notably, using 33-byte compressed keys for the silent payment address (as opposed to X-only keys in the original draft) 

It may be helpful for context to read through the discussions on #24897 , but ongoing review should happen in the relevant child PRs listed above.

w0xlt committed:
script: add `sp()`  descriptor

w0xlt committed:
wallet: add Silent Payment flag

w0xlt committed:
wallet: enable SP option in the `createwallet`

Co-authored-by: Aurèle Oulès <aurele@oules.com>

w0xlt committed:
wallet: add Silent Payment verification

Co-authored-by: josibake <josibake@protonmail.com>

w0xlt committed:
crypto: update CKey, CPubkey for silent payments

* Add methods tweaking a CKey by adding a scalar value or by multiplying
* Add a method for combining multiple CPubKeys via addition
* Add a method for doing ECDH with a CPubKey and scalar

These are the CKey,CPubkey primitives we need to implement
the silent payments protocol.

w0xlt committed:
conf: add EDCH module

josibake committed:
receiver: encode silent payment address

josibake committed:
wallet: get wallet descriptor by output type

w0xlt commented on src/wallet/test/silentpayment_tests.cpp:
```diff
@@ -0,0 +1,214 @@
+    BOOST_CHECK(EncodeSecret(silent_recipient.GetNegatedScanSecKey()) == "L2HBwB2tkUdGeb2KWx2EaUWD1YjW8u4eZ7uXJfJA8u3ibkrPtvh5");
+    BOOST_CHECK(HexStr(silent_recipient.GetScanPubkey()) == "bfa2fb9b2d094a039d4b5bf76a159d028f15a8350d5c957651d6dc00af4ccf41");
+
+    // const std::vector<std::pair<std::string, std::string>> spend_keys = {
```
These comments can be removed.
```suggestion
```

w0xlt commented on src/wallet/test/silentpayment_tests.cpp:
```diff
@@ -0,0 +1,214 @@
+
+    BOOST_CHECK(recipient_shared_secret == silent_shared_secret);
+
+    // const std::vector<std::pair<std::string, std::string>> results = {
```
These comments can be removed.

```suggestion
```

w0xlt reviewed (commented):
Approach ACK naturally,
Will review in detail soon.

Thanks for moving this project forward and splitting the draft PR in smaller ones.

luke-jr commented:
> Adds support to the Bitcoin Core wallet for receiving silent payments

Prefer to leave this for a second PR

josibake commented:
> > Adds support to the Bitcoin Core wallet for receiving silent payments
> 
> Prefer to leave this for a second PR

Can you explain your reasoning for wanting to split it up? I'd prefer to keep them together only because I see no reason to merge sending without receiving support in Bitcoin Core. In terms of work, if the goal is to merge both sending and receiving support, the amount of review doesn't change, and splitting into two PRs seems like unnecessary shuffling of commits

dergoegge commented on src/primitives/transaction.h:
```diff
@@ -159,20 +159,22 @@ class CTxOut
 public:
     CAmount nValue;
     CScript scriptPubKey;
+    bool m_silentpayment;
```
In my opinion, this flag doesn't belong here because only the receiver/sender of a silent payment would care about it, which would make it a piece of wallet meta-data.
 
Without knowing much about the wallet my guess would be that it should be stored on `CWalletTx`?

dergoegge commented on src/pubkey.h:
```diff
@@ -271,6 +281,9 @@ class XOnlyPubKey
     std::optional<std::pair<XOnlyPubKey, bool>> CreateTapTweak(const uint256* merkle_root) const;
 
+    /** Tweak an x-only public key by adding the generator multiplied with tweak32 to it. */
+    XOnlyPubKey AddTweak(const unsigned char *tweak32) const;
```
Maybe make this a uint256&? Passing a raw pointer without a size (expecting it to be 32 byte) seems a little bug prone.
```suggestion
    XOnlyPubKey AddTweak(const uint256& tweak32) const;
```

dergoegge commented on src/key.h:
```diff
@@ -101,6 +101,20 @@ class CKey
 
+    //! Tweak a secret key by adding a scalar value to it.
+    //
+    CKey AddTweak(const unsigned char *tweak32) const;
+
+    //! Tweak a secret key by multiplying it by a scalar value.
+    CKey MultiplyTweak(const unsigned char *tweak32) const;
```
Same as other suggestion
```suggestion
    CKey AddTweak(const uint256& tweak32) const;

    //! Tweak a secret key by multiplying it by a scalar value.
    CKey MultiplyTweak(const uint256& tweak32) const;
```

josibake committed:
Add BIP352 test vectors as unit tests

Use the test vectors to test sending and receiving. A few cases are not
covered here, namely anything that requires testing specific to the
wallet. For example:

* Taproot script path spending is not tested, as that is better tested in
  a wallets coin selection / signing logic
* Re-computing outputs during RBF is not tested, as that is better
  tested in a wallets RBF logic

The unit tests are written in such a way that adding new test cases is
as easy as updating the JSON file.

Also adds a comparator to `CRecipient`, to make testing easier.

josibake committed:
Increase the Bech32m limit if decoding SP address

Bech32m imposes a 90 character limit when decoding strings. Since a
silent payment address is the concatenation of two public keys, raise
this limit to 1024 when decoding a silent payment address.

The new limit is much higher than needed to account for forward
compatibility: new silent payment versions may add new data to the data
part of the address.

josibake committed:
Add "sp" HRP

josibake committed:
Add functions for decoding SP addresses

Add a function for decoding the string address and a second
function for decoding the data part of the silent payment address.

josibake committed:
Implement BIP352: Silent Payments

Implement sending and receiving, per BIP352. This is done without
requiring a full wallet, in order to simplify unit testing and to create
a more clear boundary as to what pertains to the BIP and what is left to
the wallet to decided as an implementation.

To correctly hash the public keys, per the BIP, update the HashWriter.write
method to accept a Span<unsigned char>. This allows us to easily pass a
public key in for hashing, where only the data from the public key is hashed.

This commit also moves the `CRecipient` struct to wallet/types.h and
adds a new struct, `V0SilentPaymentDestination`. These are added here to
simplify implementing sending and receiving in the future.

josibake committed:
Add a function for encoding an SP address

josibake committed:
wallet: make coin selection silent payment aware

Add a flag to the `CoinControl` object if silent payment destinations
are provided.

During coin selection, if this flag is set, skip taproot inputs when
script spend data is available. This is based on the assumption that if
a user provides script spend data, they don't have access to the key
path spend. As future improvement, we could instead check to see if we
have access to the key path spend, and only exclude the output when we
don't regardless of whether or not the user provides script spend data.

Also skip UTXOs of type `WITNESS_UNKNOWN`, although it is very unlikely
our wallet would ever try to spend a witness unknown output.

josibake committed:
wallet: add method for retreiving a private key

Add a method for retreiving a private key for a given scriptPubKey.
If the scriptPubKey is a taproot output, tweak the private key with the
merkle root or hash of the public key, if applicable.

josibake committed:
crypto: add method for applying the taptweak

The wallet returns an untweaked internal key for taproot outputs. If the
output commits to a tree of scripts, this key needs to be tweaked with
the merkle root. Even if the output does not commit to a tree of
scripts, BIP341/342 recommend commiting to a hash of the public key.

Depending how the taproot output was created, we need to apply the merkle
root tweak or hash of the public key tweak before doing ECDH

josibake committed:
wallet: create new type OutputType::SILENT_PAYMENT

josibake committed:
refactor, wallet: use std::variant<CRecipient>

Instead of using CRecipient in the transaction creation workflow, use
std::variant<CRecipient>.

This sets us up to add V0SilentPaymentDestination as a new destination in a future commit.

josibake committed:
wallet: extend Destination variant

Add V0SilentPaymentDestination to the Destination variant. This commit
does not add any logic for what to do with the new destination type, so
this commit shouldn't introduce any behavior changes.

brunoerg commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -69,6 +73,7 @@ static RPCHelpMan getwalletinfo()
                         {RPCResult::Type::BOOL, "descriptors", "whether this wallet uses descriptors for scriptPubKey management"},
                         {RPCResult::Type::BOOL, "external_signer", "whether this wallet is configured to use an external signer such as a hardware wallet"},
                         {RPCResult::Type::BOOL, "blank", "Whether this wallet intentionally does not contain any keys, scripts, or descriptors"},
+                        {RPCResult::Type::BOOL, "silent_payment", "whether this supports silent payments"},
```
```suggestion
                        {RPCResult::Type::BOOL, "silent_payment", "whether this wallet supports silent payments"},
```

brunoerg commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -351,6 +357,7 @@ static RPCHelpMan createwallet()
                                                                        " support for creating and opening legacy wallets will be removed in the future."},
             {"load_on_startup", RPCArg::Type::BOOL, RPCArg::Optional::OMITTED, "Save wallet name to persistent settings and load on startup. True to add wallet to startup list, false to remove, null to leave unchanged."},
             {"external_signer", RPCArg::Type::BOOL, RPCArg::Default{false}, "Use an external signer such as a hardware wallet. Requires -signer to be configured. Wallet creation will fail if keys cannot be fetched. Requires disable_private_keys and descriptors set to true."},
+            {"silent_payment", RPCArg::Type::BOOL, RPCArg::Default{false}, "Experimental. Indicates that the wallet supports Silent Payments. This means a more complex scanning logic"},
```
```suggestion
            {"silent_payment", RPCArg::Type::BOOL, RPCArg::Default{false}, "Experimental. Indicates that the wallet supports Silent Payments. This means a more complex scanning logic."},
```

josibake committed:
refactor, move-only: move GetSerializeSize call

Move the `GetSerializeSizeFromDestination` call up to the first time
loop over vecSend.

josibake committed:
move-only: add vouts after coin selection

Add the TxOuts to the transaction after coin selection. This gives us a
change to generate the silent payment outputs right after coin selection
and then add all the `CRecipients` to the transaction in one go.

There is a slight behavior change here: we are now checking if the vout
amounts are dust AFTER coin selection, whereas previously we were
checking this before coin selection.

josibake committed:
wallet: check if silent payments are possible

Add function to see if making a silent payments is possible:

* Check that the wallet has private keys
* Check that the wallet is unlocked

josibake committed:
wallet: update TransactionChangeType

If sending to a silent payment destination, the change type should be taproot

luke-jr commented:
>Can you explain your reasoning for wanting to split it up? I'd prefer to keep them together only because I see no reason to merge sending without receiving support in Bitcoin Core. 

1. The sooner sending is supported everywhere, the sooner it is *practical* for people to use it for receiving. Might as well get that ball rolling ASAP.
2. Sending support means the BIP isn't going to change under me (and break compatibility / risk coin loss) if I merge things in Knots first.
3. Core refuses to care about compatibility with other software, so I *can't* merge anything that changes the wallet format in Knots until after it's in a Core release.

josibake committed:
tests: add sending functional tests

josibake committed:
wallet: add `IsInputForSharedSecretDerivation` function

josibake committed:
wallet: add `CreateSilentPaymentOutputs` function

josibake committed:
wallet: add `SeparateDestinations` function

josibake committed:
wallet: create silent payment outputs

Using the new functions from the last few commits, put it all together:

* In `CreateTransactionInternal`, separate vecSend into a `CRecipient` vector and `V0SilentPaymentDestination` vector
* If there are silent payment destinations, call `CreateSilentPaymentOutputs` and add the newly created outputs to the `CRecipients` vector
* Convert the `CRecipients` to `CTxOut`s and add them to the transaction, same as before

josibake committed:
tests: add receiving functional test

Sjors commented:
Maybe add a quick summary in the description with the main implementation differences relative to #24897. It seems a big one is that this doesn't require an index!

You should also link from the child PRs back to this one.

josibake commented:
Closing in favor of https://github.com/bitcoin/bitcoin/issues/28536, where on-going tracking will be done.

To test, compile https://github.com/bitcoin/bitcoin/pull/28453 (has both send and receive support)
